### PR TITLE
some simplifications

### PR DESCRIPTION
**Pull request by incertia**
let us use this branch and PR to perform tasks related to #487

### CODE REVIEW

**Review by ggrieco-tob — CHANGES_REQUESTED** [review]

File: lib/Echidna/Campaign.hs [hunk 1]
```diff
@@ -137,27 +140,30 @@ shrinkGasSeq f g xs = sequence [shorten, shrunk] >>= uniform >>= ap (fmap . flip
...
+        rec   = updateGasInfo ts tseq' (insert f (used', reverse tseq') gi)
 updateGasInfo ((t, _):ts) tseq gi = updateGasInfo ts (t:tseq) gi
 
 -- | Execute a transaction, capturing the PC and codehash of each instruction executed, saving the
 -- transaction if it finds new coverage.
 execTxOptC :: (MonadState x m, Has Campaign x, Has VM x, MonadThrow m) => Tx -> m (VMResult, Int)
 execTxOptC t = do
-  og  <- hasLens . coverage <<.= mempty
-  res <- execTxWith vmExcept (usingCoverage $ pointCoverage (hasLens . coverage)) t
+  let l = hasLens . coverage
```
Comment (ggrieco-tob) [inline lib/Echidna/Campaign.hs:158]: Can we get a better name for this, instead of `l` ?

File: lib/Echidna/Campaign.hs [hunk 2]
```diff
@@ -184,26 +191,25 @@ seqMutators (c1, c2, c3) = fromList
           return . take ql $ if flp then take k gtxs ++ rtxs else take k rtxs ++ gtxs
 
 -- | Generate a new sequences of transactions, either using the corpus or with randomly created transactions
-randseq :: ( MonadRandom m, MonadReader x m
-           , Has GenDict x, Has TxConf x, Has TestConf x, Has CampaignConf x, Has Campaign x)
+randseq :: ( MonadRandom m, MonadReader x m, MonadState y m
```
Comment (ggrieco-tob) [inline lib/Echidna/Campaign.hs:194]: What is the benefit here?

File: lib/Echidna/Campaign.hs [hunk 3]
```diff
@@ -137,27 +141,30 @@ shrinkGasSeq f g xs = sequence [shorten, shrunk] >>= uniform >>= ap (fmap . flip
...
-                    _ -> updateGasInfo ts tseq' gi
+updateGasInfo ((t@(Tx (SolCall (f, _)) _ _ _ _ _ _), (_, used')):ts) tseq gi =
+  case mused of
+    Nothing -> rec
+    Just (used, _) | used' > used -> rec
+    Just (used, otseq) | (used' == used) && (length otseq > length tseq') -> rec
+    _ -> updateGasInfo ts tseq' gi
+  where mused = Data.Map.lookup f gi
+        tseq' = t:tseq
+        rec   = updateGasInfo ts tseq' (insert f (used', reverse tseq') gi)
```
[the remark this answers is not in the document]
Comment (incertia) [reply lib/Echidna/Campaign.hs:151]: not reserved just github

File: lib/Echidna/Campaign.hs [hunk 4]
```diff
@@ -137,27 +141,30 @@ shrinkGasSeq f g xs = sequence [shorten, shrunk] >>= uniform >>= ap (fmap . flip
...
 execTxOptC t = do
-  og  <- hasLens . coverage <<.= mempty
-  res <- execTxWith vmExcept (usingCoverage $ pointCoverage (hasLens . coverage)) t
+  let l = hasLens . coverage
+  og  <- l <<.= mempty
+  res <- execTxWith vmExcept (usingCoverage $ pointCoverage l) t
   let vmr = getResult $ fst res
   -- Update the coverage map with the proper binary according to the vm result
-  hasLens . coverage %= mapWithKey (\ _ s -> DS.map (\(i,_) -> (i, vmr)) s)
+  l %= mapWithKey (\_ s -> DS.map (set _2 vmr) s)
```
[the remark this answers is not in the document]
Comment (incertia) [reply lib/Echidna/Campaign.hs:163]: with lens this change is easy to make

File: lib/Echidna/Campaign.hs (hunk 2 above)
Comment (incertia) [reply lib/Echidna/Campaign.hs:194]: no explicit `runReaderT` in `callseq`

File: lib/Echidna/Campaign.hs [hunk 5]
```diff
@@ -69,8 +70,8 @@ isDone (view tests -> ts) = do
 -- | Given a 'Campaign', check if the test results should be reported as a
 -- success or a failure.
 isSuccess :: Campaign -> Bool
-isSuccess (view tests -> ts) =
-  all (\case { Passed -> True; Open _ -> True; _ -> False; }) $ snd <$> ts
+isSuccess (toListOf (tests . traverse . _2) -> ts) =
```
[the remark this answers is not in the document]
Comment (incertia) [reply lib/Echidna/Campaign.hs]: check the new version

**Review by ggrieco-tob — APPROVED** [review]

**Pull request merged**